---
title: "Breast Cancer Wisconsin"
format: html
echo: false
warning: false
message: false
---


```{r setup}
library(tidyverse)
library(janitor)
library(ggrepel)
library(ggtext)
library(stringr)
library(scales)

theme_set(
    theme_minimal() +
    theme(
    # axis.line.x.bottom = element_line(color = '#474747', linewidth = .3),
    # axis.ticks.x= element_line(color = '#474747', linewidth = .3),
    axis.line.y.left = element_line(color = '#474747', linewidth = .3),
    axis.ticks.y= element_line(color = '#474747', linewidth = .3),
    panel.grid = element_line(linewidth = .3, color = 'grey90'),
    panel.grid.major.x = element_blank(),
    panel.grid.minor = element_blank(),
    axis.ticks.length = unit(-0.15, "cm"),
    plot.background = element_blank(),
    plot.title.position = "plot",
    plot.title = element_text(family = "Ubuntu", size = 14, face = 'bold'),
    plot.subtitle = element_text(size = 10,lineheight = 1.15, margin = margin(0,0,20,0)),
    axis.title.x = element_markdown(family = "Ubuntu", hjust = .5, size = 9),
    axis.title.y = element_markdown(family = "Ubuntu", hjust = .5, size = 8),
    axis.text = element_markdown(family = "Chakra Petch", hjust = .5, size = 8, color = "#474747"),
    legend.position = "top",
    text = element_text(family = "Ubuntu"),
    plot.margin = margin(25, 25, 25, 25))

)
```

```{r load_data}
raw_data <- 
  read_tsv('2025/datasets/GCST005076.h.tsv.gz')
```

## Prepare date for manhatan plot
See https://danielroelfs.com/blog/how-i-create-manhattan-plots-using-ggplot/ and https://r-graph-gallery.com/101_Manhattan_plot.html for details.
```{r}
data2plot <- 
  raw_data |>
  janitor::clean_names() |>
  dplyr::select(chromosome, base_pair_location, p_value,rsid) |>
  mutate(log10_pvalue = log10(p_value)*-1)

data2plot <- 
  data2plot |>
  # Compute chromosome size
  group_by(chromosome) |> 
  summarise(chr_len=max(base_pair_location)) |> 
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) |>
  dplyr::select(-chr_len) |>
  # Add this info to the initial dataset
  left_join(data2plot, ., by=c("chromosome"="chromosome")) |>
  # Add a cumulative position of each SNP
  arrange(chromosome, base_pair_location)|>
  mutate( bp_cumulative=base_pair_location+tot)

axisdf <- data2plot |>
  group_by(chromosome) |>
  summarize(center=( max(bp_cumulative) + min(bp_cumulative) ) / 2 )
```

## Set highlight data
```{r}
data_highlight <-
  data2plot |>
  filter(log10_pvalue > 20)

# Get and filter gene name
library(biomaRt)
ensembl <- useEnsembl(biomart = "ENSEMBL_MART_SNP", dataset = 'hsapiens_snp')

result <- getBM(
  attributes = c("refsnp_id","associated_gene"),
  filters    = "snp_filter",
  values = data_highlight |> pull(rsid),
  mart = ensembl
)

genes_highlight <- 
  result |> 
  filter(associated_gene != '') |>
  separate_rows(associated_gene, sep = ',') |>
  filter(!str_detect(associated_gene,'LOC')) |>
  filter(!str_detect(associated_gene,'orf')) |>
  filter(!str_detect(associated_gene,'intergenic')) |>
  count(refsnp_id, associated_gene) |> 
  group_by(refsnp_id) |>
  slice_max(order_by = n, n = 1, with_ties = FALSE) |>
  dplyr::select(-n) |>
  rename(rsid = 'refsnp_id')

data_highlight <- 
  data_highlight |> left_join(genes_highlight, by = 'rsid') |>
  filter(!is.na(associated_gene)) |>
  group_by(associated_gene) |>
  slice_max(order_by = log10_pvalue, n = 1)
```

## Let's plot (finally)
```{r}
data2plot |>
  filter(log10_pvalue > 1) |>
  ggplot(aes(x = bp_cumulative, y = log10_pvalue, color = as.character(chromosome))) +
  geom_point(data = . %>% filter(log10_pvalue <= 20), size = .5) +
  geom_point(data = . %>% filter(log10_pvalue > 20), size = .5, color = "#316895") +
  annotate("rect",
           xmin = -Inf, xmax = Inf,
           ymin = 20, ymax = Inf,
           alpha = 0.4, fill = "#6FBED6") +
  geom_point(data = data_highlight, color = 'white', size = 3) +
  geom_point(data = data_highlight, color = '#EA6052', size = 2) +
  geom_text_repel(
    data = data_highlight, 
    aes(label = associated_gene), 
    family = "Ubuntu", 
    color= 'black', 
    size = 2, 
    direction = "y", hjust = "left",
    nudge_x = 30000000, 
    bg.color = 'white', 
    segment.size = 0,
    bg.r = 0.3) +
  scale_x_continuous( label = axisdf$chromosome, breaks= axisdf$center ) +
  scale_color_manual(values = rep(c("grey80", "grey70"), 22 )) +
  guides(color = 'none') +
  scale_y_continuous(limits = c(0,45),breaks = seq(0, 45, by = 5), expand = c(0,0,0,0)) +
  labs(
    y = 'â€“log<sub>10</sub>(p-value)',
    x = 'Chromosomes',
    title = 'Genome Wide Association Study: Breast Cancer',
    subtitle = str_wrap('Results from a GWAS of Breast Cancer. Each dot shows how strongly a common genetic variant (SNP) is linked to the disease. Higher values (lower p-values) means more significant (but not more effect size!).', width = 120),
    caption = 'Data from: GWAS Catalog GCST005076.\nNote: This represents a preliminary analysis and should not be interpreted as definitive evidence of association.') 
```
